### PR TITLE
feat: Created an environment that allows solving value based instances and instances where items can be rotated.

### PR DESCRIPTION
**Pull request by chouakifares**
Created an environment that allows solving value-based instances and instances where items can be rotated
- Added ExtendedBinPack env allowing for rotation of items and value-based instances.
- Added ExtendedRandomGenerator that can generate value-based instances and instances with free orientations of the items.
- Restored the BinPack env class to its state on the Jumanji master branch.

### CODE REVIEW

File: jumanji/environments/packing/bin_pack/conftest.py [hunk 1]
```diff
@@ -148,7 +148,119 @@ def __call__(self, key: chex.PRNGKey) -> State:
         )
 
 
-class DummyConstrainedGenerator(DummyGenerator):
+class DummyExtendedGenerator(DummyGenerator):
+    """Dummy instance generator used for testing. It outputs a constant instance with a 20-ft
+    container and 3 items: two identical items and a different third one to be able to
+    test item aggregation.
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/conftest.py]: Will this comment get changed automatically once you merge the support MR?

Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/conftest.py]: Update it to say the 6 rotations are allowed?

File: jumanji/environments/packing/bin_pack/conftest.py [hunk 2]
```diff
@@ -255,8 +367,13 @@ def dummy_generator() -> DummyGenerator:
 
 
 @pytest.fixture
-def dummy_constrained_generator() -> DummyGenerator:
-    return DummyConstrainedGenerator()
+def dummy_rotation_generator() -> DummyGenerator:
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/conftest.py]: Should the return type be DummyRotationGenerator?

File: jumanji/environments/packing/bin_pack/conftest.py [hunk 3]
```diff
@@ -255,8 +367,13 @@ def dummy_generator() -> DummyGenerator:
...
+    return DummyRotationGenerator()
+
+
+@pytest.fixture
+def dummy_extended_generator() -> DummyGenerator:
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/conftest.py]: Should the return type be DummyExtendedGenerator?

File: jumanji/environments/packing/bin_pack/types.py [hunk 4]
```diff
@@ -112,6 +71,57 @@ def valued_item_from_space_and_max_value(
 ItemType: TypeAlias = Union[Item, ValuedItem]
 
 
+def rotated_items_from_space(
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/types.py:74]: Is there a reason why you moved the location of this in the file? Or are there other changes made here?

File: jumanji/environments/packing/bin_pack/env.py [hunk 5]
```diff
@@ -780,23 +714,38 @@ def inclusion_check(ems: EMS, ems_mask: chex.Array) -> chex.Array:
...
         self,
+        is_rotation_allowed: bool,
+        is_value_based: bool,
         generator: Optional[Generator] = None,
         obs_num_ems: int = 40,
         reward_fn: Optional[RewardFn] = None,
         normalize_dimensions: bool = True,
         debug: bool = False,
         viewer: Optional[Viewer[State]] = None,
+        mean_value: Optional[float] = None,
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/env.py]: It might be better to rename this to `mean_item_value` and `standard_deviation_item_value` (or maybe `sd_item_value`?

File: jumanji/environments/packing/bin_pack/env.py [hunk 6]
```diff
@@ -823,30 +772,39 @@ def observation_spec(self) -> specs.Spec[Observation]:
         max_dim = max(self.generator.container_dims)
 
         if self.normalize_dimensions:
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/env.py]: It looks like this if should be removed (because it is already taken care of inside the various functions?

File: jumanji/environments/packing/bin_pack/env.py [hunk 7]
```diff
@@ -864,6 +822,112 @@ def observation_spec(self) -> specs.Spec[Observation]:
...
+        on whether features are to be normalized or not.
+
+        Args:
+            max_num_items: the maximum number of items that can be in an instance.
+            max_dim: The maximum dimension in this given instance.
+
+        Returns:
+            A dictionary with string keys of the item features and specs BoundedArray as values.
+        """
+        items_dict = self._items_dict_for_valued_items(max_num_items, max_dim)
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/env.py]: Would it be easier for this line to be replaced by _items_dict_for_rotated_items and then add the valued bit on top (then the rotated items bit below can be removed. At the moment it looks like everything in items_dict is being overwritten by what is below.

File: jumanji/environments/packing/bin_pack/env.py [hunk 8]
```diff
@@ -947,42 +1019,106 @@ def _make_observation_and_extras(
         def flatten_observation(observation: Observation) -> Observation:
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/env.py:1303]: Could you add a docstring to `_make_observation_and_extras` and `flatten_observation`?

File: jumanji/environments/packing/bin_pack/env.py [hunk 9]
```diff
@@ -947,42 +1019,106 @@ def _make_observation_and_extras(
...
+            ) = container_item = valued_item_from_space_and_max_value(
+                state.container, state.instance_max_item_value_magnitude
+            )
+        else:
+            items = cast(Item, items)
+            x_len, y_len, z_len = container_item = item_from_space(state.container)
+
+        norm_space = Space(x1=x_len, x2=x_len, y1=y_len, y2=y_len, z1=z_len, z2=z_len)
+        obs_ems = jax.tree_util.tree_map(
+            lambda ems, container: ems / container, obs_ems, norm_space
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/env.py]: To avoid confusion here and in the lambda function below could you use `normalisation_values` or something? (instead of using "container" to mean 2 different things in the 2 lambda functions)?

File: jumanji/environments/packing/bin_pack/env.py [hunk 10]
```diff
@@ -947,42 +1019,106 @@ def _make_observation_and_extras(
...
+        )
+        return obs_ems, items
+
     def _pack_item(  # type: ignore
-        self, state: State, ems_id: int, item_id: chex.Numeric, item_orientation: int
+        self,
+        state: State,
+        ems_id: int,
+        item_id: chex.Numeric,
+        item_orientation: Optional[Any] = None,
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/env.py]: Why is this `Any`?

File: jumanji/environments/packing/bin_pack/env.py [hunk 11]
```diff
@@ -947,42 +1019,106 @@ def _make_observation_and_extras(
...
-        state.items_placed = state.items_placed.at[item_orientation, item_id].set(True)
-        # jax.debug.print("items_placed : {}",state.items_placed)
+        if item_orientation is not None:
+            state.items_mask = state.items_mask.at[:, item_id].set(False)
+            state.items_placed = state.items_placed.at[item_orientation, item_id].set(
+                True
+            )
+        else:
+            state.items_mask = state.items_mask.at[item_id].set(False)
+
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/env.py]: Remove this empty space

File: jumanji/environments/packing/bin_pack/env.py [hunk 12]
```diff
@@ -1017,50 +1153,69 @@ def is_action_allowed(
...
-
-        for o in range(6):
-            tmp_items = Item(
-                state.items[:][0][o], state.items[:][1][o], state.items[:][2][o]
-            )
-            item_spaces = space_from_item_and_location(tmp_items, state.items_location)
-            ems_intersect_items = jax.vmap(Space.intersect, in_axes=(0, None))(
-                state.ems, item_spaces
+        if self.is_rotation_allowed:
+            for o in range(6):
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/env.py:1484]: I thought at some point you'd vectorised this or was that somewhere elsewhere?

File: jumanji/environments/packing/bin_pack/viewer.py [hunk 13]
```diff
@@ -318,48 +323,54 @@ def _add_overlay(self, fig: plt.Figure, ax: plt.Axes, state: State) -> None:
...
-            for j, placed in enumerate(state.items_placed[i])
-            if placed
-        )
-        return used_volume
+        if self.is_rotation_allowed:
+            used_volume = sum(
+                float(state.items.x_len[i, j])
+                * float(state.items.y_len[i, j])
+                * float(state.items.z_len[i, j])
+                for i in range(6)
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/viewer.py:370]: Wouldn't this give a used volume of 6 times the volume of the item?

File: jumanji/environments/packing/bin_pack/generator_test.py [hunk 14]
```diff
@@ -402,7 +410,7 @@ def test_random_generator__generate_solution(
         assert_trees_are_different(
             solution_state1.items_location, state1.items_location
         )
-        assert jnp.all(solution_state1.items_placed | ~solution_state1.items_mask)
+        assert jnp.all(solution_state1.items_placed[0] | ~solution_state1.items_mask)
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/generator_test.py:415]: Is this because the optimal solution is always the not rotated item? If so could you add an explanatory comment?

File: jumanji/environments/packing/bin_pack/env_test.py [hunk 15]
```diff
@@ -503,6 +535,259 @@ def test_bin_pack__optimal_policy_random_instance(
...
                 assert not jnp.any(timestep.extras["invalid_ems_from_env"])
-            assert jnp.array_equal(state.items_placed[0], solution.items_placed)
+            assert jnp.array_equal(state.items_placed, solution.items_placed)
             assert round(timestep.extras["volume_utilization"]) == 1
             assert timestep.extras["ratio_packed_items"] == 1
+
+
+class TestExtendedBinPackRotationValue:
+
+    # TO-DO:
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/env_test.py]: Are these still todo or can this comment be removed?

Could you add a docstring?

File: jumanji/environments/packing/bin_pack/generator.py [hunk 16]
```diff
@@ -1250,15 +1061,14 @@ def _generate_list_of_valued_items(
         return list_of_items
 
 
-VALUE_BASED_GENERATORS = (
-    RandomValueProblemGenerator,
-    ValueProblemCSVGenerator,
-)
+VALUE_BASED_GENERATORS = (ValueProblemCSVGenerator,)
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/generator.py]: Is this used anymore?

File: jumanji/environments/packing/bin_pack/generator.py [hunk 17]
```diff
@@ -1294,16 +1104,34 @@ def _unpack_items(self, state: State) -> State:
...
         prob_split_one_item: float = 0.7,
         split_num_same_items: int = 5,
         container_dims: Tuple[int, int, int] = TWENTY_FOOT_DIMS,
+        mean_value: Optional[float] = None,
+        standard_deviation_value: Optional[float] = None,
     ):
+        """
+        Args:
+            is_rotation_allowed: whether the generator has to generate instances where item
+            rotation is possible. Defaults to False.
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/generator.py]: indent this line

File: jumanji/environments/packing/bin_pack/generator.py [hunk 18]
```diff
@@ -1294,16 +1104,34 @@ def _unpack_items(self, state: State) -> State:
...
+            is_value_based: whether the generator has to generator has to generate
+            instance with values. Defaults to False.
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/generator.py]: indent this line

File: jumanji/environments/packing/bin_pack/generator.py [hunk 19]
```diff
@@ -1312,20 +1140,179 @@ def __init__(
...
+            )
 
-    def _generate_solved_instance(self, key: chex.PRNGKey) -> State:
+    def _generate_value_based_solved_instance(self, key: chex.PRNGKey) -> State:
+        """Generate the random instance with half of all items correctly packed (the higher value
+        half). The other half of the generated items remain unpacked.
+
+        The first half of items are generated by splitting the container in the same way as the
+        inherited RandomGenerator class, but with less than half the max_num_items. Values are
+        randomly generated, doubled and assigned to these items. These items are
```
Comment (rsjeffers) [inline jumanji/environments/packing/bin_pack/generator.py]: Comment is out of date, they aren't doubled anymore.

Is the rest of this a copy and paste of what was in the original Value based generator?

**Review by rsjeffers — COMMENTED** [review]
First review complete. Thanks for the work!

File: jumanji/environments/packing/bin_pack/types.py (hunk 4 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/types.py:74]: There are other changes (the addition of the value attribute) that required the return type to be ItemType instead of Item, I moved it down here because I needed the declaration of ItemType.

File: jumanji/environments/packing/bin_pack/generator_test.py (hunk 14 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/generator_test.py:415]: Yes, i'll add the comment.

File: jumanji/environments/packing/bin_pack/viewer.py (hunk 13 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/viewer.py:370]: it wouldn't because at most one orientation of the item would allow for placed to be True.

File: jumanji/environments/packing/bin_pack/generator.py (hunk 16 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/generator.py]: Nope, I mean the CSV generator was never used anyways. I'll remove it.

File: jumanji/environments/packing/bin_pack/generator.py (hunk 17 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/generator.py]: Done ✅

File: jumanji/environments/packing/bin_pack/generator.py (hunk 18 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/generator.py]: Done ✅

File: jumanji/environments/packing/bin_pack/generator.py (hunk 19 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/generator.py]: They're still doubled. And Yes I copy-pasted the whole function from your previous generator.

File: jumanji/environments/packing/bin_pack/env.py (hunk 10 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env.py]: I changed it to `int`.

File: jumanji/environments/packing/bin_pack/env.py (hunk 11 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env.py]: Done ✅

File: jumanji/environments/packing/bin_pack/env.py (hunk 5 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env.py]: Done ✅

File: jumanji/environments/packing/bin_pack/env.py (hunk 6 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env.py]: Done ✅

File: jumanji/environments/packing/bin_pack/env.py (hunk 7 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env.py]: You're right. I added it.

File: jumanji/environments/packing/bin_pack/env.py (hunk 8 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env.py:1303]: Done ✅

File: jumanji/environments/packing/bin_pack/env.py (hunk 9 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env.py]: Done ✅

File: jumanji/environments/packing/bin_pack/env.py (hunk 12 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env.py:1484]: I didn't vectorize this, and I didn't think it was worth vectorizing since it is only used in debug mode

File: jumanji/environments/packing/bin_pack/conftest.py (hunk 1 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/conftest.py]: I will update the comment, but when merging the support MR , this comment will stay the same since it is correct in this case, I removed the bit about item aggregation though.

File: jumanji/environments/packing/bin_pack/conftest.py (hunk 2 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/conftest.py]: Done ✅

File: jumanji/environments/packing/bin_pack/conftest.py (hunk 3 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/conftest.py]: Done ✅

File: jumanji/environments/packing/bin_pack/env.py (hunk 8 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/env.py:1303]: The _make_observation_and_extras is missing Returns in the docstring

File: jumanji/environments/packing/bin_pack/env.py (hunk 9 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/env.py]: Actually, could you change it to normalisation_ems and normalisation_item so that it isn't confused with the item values?

File: jumanji/environments/packing/bin_pack/generator.py (hunk 16 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/generator.py]: Yes the CSV generator can be used there wasn't a test for it I believe? I think this list (with all value based generators) was there to make sure that ValuedItems were set instead of Items. Could you write a test to check this? We will need this generator for the inference runs. Thanks!

File: jumanji/environments/packing/bin_pack/generator.py (hunk 19 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/generator.py]: The values are not doubled, the items are duplicated. The values are value + total value of original items (though it needs to be explained clearer than this).

File: jumanji/environments/packing/bin_pack/viewer.py (hunk 13 above)
Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/viewer.py:370]: ah yes I missed the if. Thanks!

File: jumanji/environments/packing/bin_pack/env.py (hunk 8 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env.py:1303]: Done ✅

File: jumanji/environments/packing/bin_pack/env.py (hunk 9 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env.py]: done ✅

File: jumanji/environments/packing/bin_pack/generator_test.py (hunk 14 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/generator_test.py:415]: Done ✅

File: jumanji/environments/packing/bin_pack/env_test.py (hunk 15 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/env_test.py]: Done ✅

File: jumanji/environments/packing/bin_pack/generator.py (hunk 19 above)
Comment (chouakifares) [reply jumanji/environments/packing/bin_pack/generator.py]: I modified the comment but feel free to add a suggestion here.

Comment (rsjeffers) [reply jumanji/environments/packing/bin_pack/generator.py]: Great thanks!

**Review by rsjeffers — APPROVED** [review]
Ok all threads resolved!
